server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadataStats.java

```diff
@@ -0,0 +1,164 @@
+            final var shardStats = Arrays.stream(indexShardsStats.getShards())
+                .filter(stats -> stats.getShardRouting().primary())
+                .findFirst()
+                // Fallback to a replica if for some reason we couldn't find the primary stats
+                .orElse(indexShardsStats.getAt(0));
+            final var commonStats = shardStats.getStats();
+            final var indexingShardStats = commonStats.getIndexing().getTotal();
+            indexWriteLoadBuilder.withShardWriteLoad(
+                shardStats.getShardRouting().id(),
+                indexingShardStats.getWriteLoad(),
+                indexingShardStats.getTotalActiveTimeInMillis()
+            );
+            totalSizeInBytes += commonStats.getStore().sizeInBytes();
+            shardsTookIntoAccountForSizeAvg++;
+        }
+
+        return new IndexMetadataStats(
+            indexWriteLoadBuilder.build(),
+            new AverageShardSize(totalSizeInBytes, shardsTookIntoAccountForSizeAvg)
+        );
+    }
+
+    public IndexWriteLoad writeLoad() {
+        return indexWriteLoad;
+    }
+
+    public record AverageShardSize(long totalSizeInBytes, int numberOfShards) implements Writeable, ToXContentFragment {
+
+        public static final ParseField TOTAL_SIZE_IN_BYTES_FIELD = new ParseField("total_size_in_bytes");
+        public static final ParseField SHARD_COUNT_FIELD = new ParseField("shard_count");
+
+        @SuppressWarnings("unchecked")
+        private static final ConstructingObjectParser<AverageShardSize, Void> PARSER = new ConstructingObjectParser<>(
+            "average_shard_size",
+            false,
+            (args, unused) -> new AverageShardSize((long) args[0], (int) args[1])
+        );
+
+        static {
+            PARSER.declareLong(ConstructingObjectParser.constructorArg(), TOTAL_SIZE_IN_BYTES_FIELD);
+            PARSER.declareInt(ConstructingObjectParser.constructorArg(), SHARD_COUNT_FIELD);
+        }
+
+        AverageShardSize(StreamInput in) throws IOException {
+            this(in.readLong(), in.readInt());
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            out.writeLong(totalSizeInBytes);
+            out.writeInt(numberOfShards);
+        }
+
+        @Override
+        public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+            builder.field(TOTAL_SIZE_IN_BYTES_FIELD.getPreferredName(), totalSizeInBytes);
+            builder.field(SHARD_COUNT_FIELD.getPreferredName(), numberOfShards);
+            return builder;
+        }
+
+        static AverageShardSize fromXContent(XContentParser parser) throws IOException {
+            return PARSER.parse(parser, null);
+        }
+
+        public OptionalLong getAverageSizeInBytes() {
+            long value = totalSizeInBytes / numberOfShards;
+            return numberOfShards == 0 ? OptionalLong.empty() : OptionalLong.of(value);
```

Comment:
This looks like a division by zero opportunity? I think we need to inline the value calculation instead.

Reply:
Right 🤦

Reply:
I guess `numberOfShards` is never 0 here? I found no tests checking that `getAverageShardSize().isPresent == false`. Perhaps we can remove the `OptionalLong` usage here?

Reply:
I added an extra test in https://github.com/elastic/elasticsearch/pull/91561/commits/6b7488d5477fcd31ba12ebf7246b1fd6525ef10e to cover that case.

Reply:
I wonder if that test provokes the `numberOfShards==0` case? If there are no stats, I'd expect the IndexMetadataStats object to be null instead?

Reply:
We return a non-null `IndexMetadataStats` in `IndexMetadataStats#fromStatsResponse` in that case. Maybe we want to return null when there are no stats available?